### DISCUSSION TITLE
(Unexpected) Exception and Error Handling: UX

### DISCUSSION THREAD

**Discussion by ViMaSter** [Studio-Lovelies/GG-JointJustice-Unity#198] [Ideas]
Branching off of https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/189#discussion_r810544860...

Reviewing this PR I realized that even in (non-[Development )Builds](https://docs.unity3d.com/Manual/BuildSettings.html#:~:text=Function-,Development%20Build,-A%20development%20build) of Unity Games, **unhandled exceptions will not terminate the game**.  
Unity games are compiled to native machine code and thereby don't utilize the Common Language Runtime and thereby Infrastructure.
**This is different to how [CLI C#](https://en.wikipedia.org/wiki/Common_Language_Infrastructure) apps [react to unhandled exceptions](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/exceptions/#:~:text=When%20an%20exception,to%20the%20user.).**

From my personal experience working with C# unexpected exceptions _will_ happen. This can have a variety of reasons:
- downloaded builds are incomplete
  - either because our build process forgot to upload a file
  - the `.zip` downloaded by players was incomplete but still unpacked
- a user can run out of storage while playing the game, which will result in us failing to create save files
- the DirectX or OpenGL installation on the users machine can be faulty
- <sup>a user might use more than 255 keyboards 👀</sup>

All of these are things **we should handle**, but not in a way that allows the user to return to the game as usual:
- if downloaded builds are incomplete, scripts might load incorrectly or music/sprites are missing and the game is effectively corrupted
- if the user attempts to save but ran out of storage (because of things like background Steam downloads), we shouldn't just "not save" their settings and only have the user discover it the next time the starts the game
- [if the user's GPU decides to shuffle its memory](https://www.techpowerup.com/forums/attachments/img_1979-jpg.138574/) (this could be driver or hardware issues), we don't want to continue loading in assets
- <sup>~~if a user plugs in the 255 keyboards there's clearly bigger issues at hand and I'm pretty sure it's beyond saving anyway~~</sup>

---

[There's also this thread about **the code-side of things**.](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/discussions/199)

---

# User experience for unhandled exceptions

My suggestion UX-wise would be pretty straight forward:
- We set up a simple scene with a user-friendly text, that the game has unrecoverably crashed, needs to be restarted (and to redownload it, if the error persists) and a button to quit the game.
- A small note: **Web Player builds can't exit**, as they run in a browser we have no control over. Here I would simply hide the button and instead ask the user to reload the page (ideally with `Ctrl+Shift+F5`, so the cache is cleared too).

**Comment by notexplosive** [comment]
I like the UX you're proposing but am concerned about implementation.

- How easy is it to make any uncaught exception lead to a "you crashed" scene? (maybe this is more C# wizardry I'm not aware of)
- What if we throw an exception while loading the crash scene?

**Reply by ViMaSter** [reply]
> How easy is it to make any uncaught exception lead to a "you crashed" scene? (maybe this is more C# wizardry I'm not aware of)

Was making assumptions [based on regular CLI/CLR apps (`AppDomain.CurrentDomain.UnhandledException += CoolEvent;`)](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.unhandledexception?view=net-6.0#:~:text=new%20UnhandledExceptionEventHandler(MyHandler)%3B) where you have a global event you can hook into, but turns out, Unity actually [supports a similar flow except it's for the logs of the console and you'll have to filter for exceptions](https://stackoverflow.com/a/41564034/1088653).  
As long as we have the game start in a scene where a Component like that is constructed and then make sure it doesn't get destroyed between loading new scenes, we can just have that as a failsafe that's always active in the background.

> What if we throw an exception while loading the crash scene?

We'd technically land in that same handler, so we'd have to keep track of whether or not we're in the "recovery loop" and at that point probably can only call [`Application.Quit(errorCode)`](https://docs.unity3d.com/ScriptReference/Application.Quit.html). From what I've seen, Unity has no built-in way to present a modal dialog or anything native.  
Though if we really wanted to, we can always build platform-specific implementations and just use compiler flags to call the appropriate library based on what target platform we're building for.
